Expose Thanos HTTP APIs to OpenAPI/protobuf Proposal (#4426)
thanos-io/thanos: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md

> | type     | title                                                        | status   | menu               |

Reviewer: I think we use different format of front matter - hopefully we can get rid of this too, but right now it's required to be yaml between `---` letters.

| -------- | ------------------------------------------------------------ | -------- | ------------------ |
| proposal | Expose Thanos APIs to OpenAPI/protobuf and expose on website and UIs | accepted | proposals-accepted |

## **Expose Thanos APIs to OpenAPI/protobuf and expose on website and UIs**

- Owners:
  - [@Hangzhi](https://github.com/Hangzhi)
- Related Tickets:
  - Program Proposal: https://github.com/thanos-io/thanos/issues/4102
  - Need for auto-documentation in Prometheus:
    - https://github.com/prometheus/prometheus/issues/7192
    - https://github.com/prometheus/prometheus/issues/5567
- Other docs:
  - A proposal of LFX mentorship: https://docs.google.com/document/d/1GeyFCk4QkVSrx4pQ4OSmOA1KE_FvmJRi_aq9hr2GoZQ/edit?usp=sharing

This design doc is proposing a design for Thanos HTTP API defined in protobuf and OpenAPI.

## **Motivations**

> In order to improve Thanos usage for users, we would like to define our HTTP APIs in protobuf/OpenAPI and expose those in the website. This would allow users to use tools for documentation, validation, type checking, and even code generation to use our APIs efficiently.

Reviewer: ```suggestion
In order to improve Thanos usage for users, we would like to define our HTTP APIs in protobuf/OpenAPI and expose those in our repository. This would allow users to use tools for documentation, validation, type checking, and even code generation to use our APIs efficiently.

Similarly we want reuse this work in Prometheus.

```

Reviewer 2: For readability it would be good to expand on the following parts:
* `In order to improve Thanos usage for users` - what do we mean by improve here? What positive benefit will this bring to users?  
* `protobuf/OpenAPI` - as far as I am aware these are two different & non-interchangeable things?

### **Pitfalls of the current solution**

> - Documentation needs to be written manually.

Reviewer: ```suggestion
- Documentation, Server code and client code boilerplate needs to be written manually.
- It's hard to discover the current API programmatically.
- It's easy to make an accidental breaking changes when modifying API
```

Reviewer 2: I would reword the final point to be something like
> When modifying the API, it is very difficult to know if this breaks downstream users

Or similar :)

## **Goals**

- Define all APIs in protobuf.
- Be able to generate OpenAPI3 from protobuf.
- Auto-generate documentation with OpenAPI3 specification.
- Generate server code from API specification (OpenAPI3 or protobuf).

> - Define all configuration potentially in protobuf too: https://github.com/openproto/protoconfig.

Reviewer: Maybe we can drop that to focus more on APIs in this phase (:

> - Define all configuration potentially in protobuf too: https://github.com/openproto/protoconfig.

Reviewer 2: IMO these are not goals _per-se_ but the steps we take to satisfy the goals - in fact, they are very similar to the `How` section below.

I think the goals here is to make the Thanos APIs easier to share, document and consume?

- Optionally we would like to have them on the index page on every Thanos component server.

## **Non-Goals**

> - Not define Thanos gRPC APIs in protobuf specification.

Reviewer: ```suggestion
- Don't mix gRPC with HTTP APIs in same protobuf package

```

## **How**

> - Define Thanos APIs in protobuf specification.

Reviewer: ```suggestion
- Define Thanos HTTP APIs in protobuf specification.

```

- Generate OpenAPI from protobuf with gnostic extension.
- Generate documentation from OpenAPI with [swagger](https://github.com/swagger-api/swagger-codegen).

> - Generate server API client and server stubs potentially from OpenAPI with [swagger](https://github.com/swagger-api/swagger-codegen) or [oapi-codegen](https://github.com/deepmap/oapi-codegen).

Reviewer: ```suggestion
- Generate server and client API stubs potentially from OpenAPI with [swagger](https://github.com/swagger-api/swagger-codegen) or [oapi-codegen](https://github.com/deepmap/oapi-codegen).

```

Reviewer 2: I'm not a big fan of having 'potentially' in proposals - we're either going to do it or not :)

## **Alternatives**

> ## **Only define HTTP RESTful API only in OpenAPI. Not in protobuf.**

Reviewer: ```suggestion
## **Define HTTP RESTful API only in OpenAPI. Not in protobuf.**

```

1. Pros:

>    1. We are not sure whether we can definitely generate OpenAPI specifications from protobuf.

Reviewer: * There are might be some complexity, edge case and extra tooling to make 3 step process (proto -> open API -> documentation to work

2. Cons:
   1. We write API definitions in protobuf for consistency.
   2. Protobuf definition is more concise.

## **Define APIs in gRPC and have RESTful APIs alongside with [grpc - gateway](https://github.com/grpc-ecosystem/grpc-gateway)**

1. Pros:
   1. We can have gRPC and RESTful APIs at the same time.
   2. We have gRPC APIs like rules API.
2. Cons:

>    1. We don't know clearly how grpc-gateway works.

Reviewer: * We need run another sidecar (complexity of running the system)
* Semantics of gRPC and HTTP might be different and surprising for end user.
* We want to reuse in Prometheus and Prometheus does not support gRPC (gRPC dependency was removed from codebase)
* Same port library is not maintained. (cmux)

So, we want to define REST APIs in protobuf, generate OpenAPI definition from protobuf with [gnostic](https://github.com/google/gnostic). This would allow users to use tools for documentation, validation, type checking, and even interface code generation to use our APIs efficiently.

Similarly, we want to reuse this work in Prometheus.

### **Pitfalls of the current solution**

* Documentation, Server code, and client code boilerplate needs to be written manually.
* Hand-written interface code is expensive to write and hard to maintain. 
* It's hard to discover the current API programmatically.
* When modifying the API, it is very difficult to know if this breaks downstream users. 

## **Goals**

* Define all APIs in protobuf.
* Be able to generate OpenAPI3 from protobuf.
* Auto-generate documentation with OpenAPI3 specification.
* Generate server code from API specification (OpenAPI3 or protobuf).

> * Define all configuration potentially in protobuf too: https://github.com/openproto/protoconfig.

Reviewer 3: I don't think this goal is in the scope of this proposal, let's remove this.

   1. There are might be some complexity, edge case and extra tooling to make the 3-step process (proto -> OpenAPI -> documentation to work.
    
2. Cons:
   1. We write API definitions in protobuf for consistency. OpenAPI is less consistent compared to protobuf in a project built on Golang. 
   2. Protobuf definition is more concise to write. 

## **Define APIs in gRPC and have RESTful APIs alongside with grpc - gateway**

1. Pros:
   1. We can have gRPC and RESTful APIs at the same time.
   2. We have gRPC APIs like rules API.
2. Cons:
   1. There are might be some complexity, edge cases and extra tooling to make the process (define gRPC API and get RESTful APIs with grpc-gateway) work. 
   2. We need to redefine our API service in gPRC since most of Thanos APIs and Prometheus APIs are RESTful APIs. 
   3. We need to run another sidecar (complexity of running the system).
   4. Semantics of gRPC and HTTP might be different and surprising for end user. 
   5. We want to reuse in Prometheus and Prometheus does not support gRPC (gRPC dependency was removed from codebase).
   6. Same port library is not maintained. (cmux)

## **Action Plan**

* [ ]  Experiments in rules API.

> * [ ]  Define all http APIs in Protobuf.

Reviewer 3: ```suggestion
* [ ]  Define all HTTP APIs in Protobuf.

```

> * [ ]  Generate documents, client code, and server stubs from OpenAPI specification.

Reviewer 3: ```suggestion
* [ ]  Generate documentation, client code, and server stubs from OpenAPI specification.

```
